README.md
# fetch-again

[![npm version](https://badge.fury.io/js/fetch-again.svg)](https://badge.fury.io/js/fetch-again)

A tiny, dependency-free TS library for adding retry logic to your favorite fetch implementation.

## HERE BE DRAGONS

This library has not been tested for production use.

## Installation

Using npm:

```bash
npm install @helmturner/fetch-again
```

Using yarn:

```bash
yarn add @helmturner/fetch-again
```

## Usage

```typescript
import fetchWithRetry from "@helmturner/fetch-again";

const response = await fetchWithRetry(options, "https://example.com", {
  method: "POST",
  body: JSON.stringify({ foo: "bar" }),
});
```

## API

### `fetchWithRetry(options, ...fetchArguments)` -> `Promise<Response>` | `Promise<Error>`

#### `options`

##### `fetchFn` (optional)

A fetch implementation to use. If not provided, the library will attempt to find a fetch implementation in the following order:

1. `fetch` (global)
2. `window.fetch` (browser)
3. `global.fetch` (node)
4. `require("node-fetch")` (must be installed as a dependency)

##### `retries` (optional)

The number of times to retry the request. Defaults to `3`.

##### `factor` (optional)

The exponential factor to use. Defaults to `2`. The timeout between retries will grow by this factor between each retry, e.g. for a factor of `2` the timeout will double with each retry.

##### `minTimeout` (optional)

The minimum timeout between retries, in milliseconds. Defaults to `1000`.

##### `maxTimeout` (optional)

The maximum timeout between retries, in milliseconds. Defaults to `10000`.

##### `randomize` (optional)

Whether to randomize the timeout between retries. Defaults to `false`.

##### `onRetry` (optional)

A callback to be called after each retry. The callback will be passed the result of the request (regardless of whether returned or thrown) and the number of retries remaining. Defaults to a no-op.

##### `retryOn` (optional)

A function that determines whether to retry the request. The function will be passed the result of the request (regardless of whether it was returned or thrown) and the number of retries remaining.

By default, the following steps are taken to determine whether to retry:

1. If the number of retries remaining is `0`, the request will not be retried.
2. If the result is an instance of `Error` or `null`, the request will be retried.
3. If the result has an `ok` property, the request will only be retried if the value of `ok` is **falsey**
4. Otherwise, If the result has any property that is a number between `100` and `599` inclusive, the request will be retried only if the value of that property is `408`, `413`, `429`, `500`, `502`, `503` or `504`.
5. If none of the above conditions are met, an error will be thrown.

**_IT IS HIGHLY RECOMMENDED THAT YOU PROVIDE A CUSTOM `retryOn` FUNCTION IF USING A FETCH IMPLEMENTATION OTHER THAN `node-fetch` OR THE `fetch` API_**

#### `...fetchArguments`

The arguments to pass to the fetch implementation. The arguments should be the same as the arguments you would pass to the fetch implementation you are using - TypeScript will type them according to the fetch function passed.

## Examples

With default `fetch` implementation:

```typescript
import fetchWithRetry from "@helmturner/fetch-again";

const options = {
  retries: 5,
  factor: 2,
  minTimeout: 1000,
  maxTimeout: 10000,
  randomize: true,
  onRetry: (result, count) => {
    console.log(`Retrying request. ${count} retries remaining.`);
  },
  retryOn: (result, count) => {
    if (count === 0) return false;
    if (result instanceof Error) return true;
    if (result.status === 429) return true;
    return false;
  },
};

const response = await fetchWithRetry(options, "https://example.com", {
  method: "POST",
  body: JSON.stringify({ foo: "bar" }),
});
```

With custom `fetch` implementation:

```typescript
import fetchWithRetry from "@helmturner/fetch-again";
import axios from "axios";

const options = {
  fetchFn: axios,
  retries: 5,
  factor: 2,
  minTimeout: 1000,
  maxTimeout: 10000,
  randomize: true,
  onRetry: (result, count) => {
    console.log(`Retrying request. ${count} retries remaining.`);
  },
  retryOn: (result, count) => {
    if (count === 0) return false;
    if (error.response && typeof error.response.status === "number") {
      return [408, 413, 429, 500, 502, 503, 504].includes(
        error.response.status
      );
    }
    return true;
  },
};

const response = await fetchWithRetry(options, {
    url: "https://example.com",
    method: "get",
});
```

## License

ISC License (ISC)
Copyright (c) 2023 Helm Turner


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/alecvision/fetch-again/compare/v1.0.1...v1.0.2) (2023-02-05)

### 1.0.1 (2023-02-05)


### Bug Fixes

* **(errors,types):** throw when `fetchFn` passed without `retryOn`; Improve types; ([5d8c49e](https://github.com/alecvision/fetch-again/commit/5d8c49ead31f37355d3274aaaeec3be1b21d9d12))


src/index.test.ts
import { describe, it, expect, TestOptions } from "vitest";
import fetchWithRetry, { RetryOptions } from ".";

const settings: TestOptions = { timeout: 30000 };

describe.concurrent(
  "fetchWithRetryCompatibilityTesting",
  async () => {
    it("should retry 408 when used with the default fetch", async () => {
      const counter = { count: 0 };

      const response = await fetchWithRetry(
        {
          retries: 3,
          onRetry: (result, count) => {
            counter.count++;
          },
        },
        "https://httpstat.us/408",
        {
          method: "GET",
        }
      );
      expect(response).not.toBeInstanceOf(Error);
      if (response instanceof Error) return;
      expect(response.status).toBe(408);
      expect(counter.count).toBe(3);
      expect(response.ok).toBe(false);
    });

    it("should retry 413 when used with the default fetch", async () => {
      const counter = { count: 0 };
      const config: RetryOptions = {
        retries: 3,
        onRetry: (result, count) => {
          counter.count++;
        },
      };

      const response = await fetchWithRetry(config, "https://httpstat.us/413", {
        method: "GET",
      });

      expect(response.status).toBe(413);
      expect(counter.count).toBe(3);
      expect(response.ok).toBe(false);
    });

    it("should throw if a 3rd-party fetcher is supplied without retryOn also being supplied", async () => {
      const axios = await import("axios").then((module) => module.default);
      const counter = { count: 0 };
      // @ts-expect-error
      const config: RetryOptions<
        typeof axios.get<any>,
        import("axios").AxiosError
      > = {
        fetchFn: axios.get,
        retries: 3,
        onRetry: (result, count) => {
          counter.count++;
        },
      };

      expect(
        fetchWithRetry(config, "https://httpstat.us/413", {
          method: "GET",
        })
      ).rejects.toThrow();
    });
  },
  settings
);

//408, 413, 429, 500, 502, 503, 504


src/index.ts




/**
 * The options object for the fetchWithRetry function.
 * @template TFetchFn The type of the function passed as fetchFn. If not
 * provided, the global fetch function will be used. If the global fetch
 * is not available, node-fetch will be used, if installed.
 * @template TErrorType The type of the error that will be thrown if the
 * request cannot be retried. Defaults to the return type of the fetchFn.
 *
 * @example (using axios)
 * ```typescript
 * const options: RetryOptions<typeof axios.get<SomeRemoteAPIData>, AxiosError> = {
 *   fetchFn: axios.get,
 *   retries: 3,
 *   factor: 2,
 *   minTimeout: 1000,
 *   maxTimeout: 10_000,
 *   randomize: false,
 *   retryOn: (result, count) => {
 *     // `result` is typed as either TErrorType or the return type of TFetchFn here.
 *   },
 *   onRetry: (result, count) => {
 *     // `result` is typed as either TErrorType or the return type of TFetchFn here.
 *   }
 * ```
 */

export type RetryOptions<
  TFetchFn extends
    | ((...arguments_: any[]) => Promise<any>)
    | undefined = undefined,
  TErrorType = TFetchFn extends undefined
    ? ResultOf<DefaultFetch>
    : ResultOf<TFetchFn>
> = Omit<
  Parameters<
    TFetchFn extends undefined
      ? typeof fetchWithRetry<
          Parameters<DefaultFetch>,
          ResultOf<DefaultFetch>,
          DefaultFetch,
          TErrorType
        >
      : typeof fetchWithRetry<
          Parameters<TFetchFn>,
          ResultOf<TFetchFn>,
          TFetchFn,
          TErrorType
        >
  >[0],
  "fetchFn" | "retryOn"
> &
  (TFetchFn extends undefined
    ? {
        fetchFn?: undefined;
        retryOn?: (result: ResultOf<DefaultFetch>, count: number) => boolean;
      }
    : {
        fetchFn: TFetchFn;
        retryOn: (result: ResultOf<TFetchFn>, count: number) => boolean;
      });

type DefaultFetch = ResultOf<typeof getDefaultFetchImplementation>;
type ResultOf<T> = T extends (...args: any[]) => Promise<infer P>
  ? P
  : T extends (...args: any[]) => infer R
  ? R
  : never;

const defaultRetryTest = (result: unknown, count: number) => {
  if (count === 0) return false;

  if (typeof result !== "object") {
    throw new TypeError(
      `Unexpected result type: Expected object, got ${typeof result}`
    );
  }

  if (result === null) return true;
  if ("ok" in result) return !result.ok;

  const returnCode = Object.values(result).find((value): value is number => {
    return typeof value === "number" && value > 99 && value < 600;
  });

  if (returnCode) {
    return new Set([408, 413, 429, 500, 502, 503, 504]).has(returnCode);
  }

  throw new TypeError(
    "Unable to determine if request can be retried. " +
      "If you are manually providing a fetch function, " +
      "you must also provide a retryOn function."
  );
};

async function getDefaultFetchImplementation() {
  if (typeof fetch !== "undefined") return fetch;
  if (typeof window !== "undefined" && window.fetch) return window.fetch;
  if (typeof global !== "undefined" && global.fetch) return global.fetch;
  try {
    const { default: nodeFetch } = await import("node-fetch");
    return nodeFetch;
  } catch {
    throw new Error(
      "Unable to find fetch implementation. " +
        "Either pass `fetchFn` and `retryOn` in the options object " +
        "or install `node-fetch` as a dependency."
    );
  }
}

export default async function fetchWithRetry<
  TFetchParams extends any[] = any[],
  TReturn = ResultOf<DefaultFetch>,
  TCustomFetchFn extends
    | ((any: any) => Promise<TReturn>)
    | undefined = undefined,
  TError = TReturn
>(
  options: (undefined extends TCustomFetchFn
    ? { retryOn?: (result: TReturn | TError, count: number) => boolean }
    : { retryOn: (result: TReturn | TError, count: number) => boolean }) & {
    fetchFn?: TCustomFetchFn | undefined;
    retries?: number | undefined;
    factor?: number | undefined;
    minTimeout?: number | undefined;
    maxTimeout?: number | undefined;
    randomize?: boolean | undefined;
    onRetry?: ((result: TReturn | TError, count: number) => void) | undefined;
  },
  ...fetchArguments: TFetchParams
): Promise<TReturn | TError> {
  if (options.fetchFn && !options.retryOn) {
    throw new TypeError(
      "If you are manually providing a fetch function, " +
        "you must also provide a retryOn function. " +
        "If you wish to use the global fetch or node-fetch, " +
        "do not provide a fetchFn; whichever is available will be used."
    );
  }

  const config = {
    fetchFn: options.fetchFn ?? (await getDefaultFetchImplementation()),
    retries: options.retries ?? 3,
    factor: options.factor ?? 2,
    minTimeout: options.minTimeout ?? 1000,
    maxTimeout: options.maxTimeout ?? 10_000,
    randomize: options.randomize ?? false,
    retryOn: options.retryOn ?? defaultRetryTest,
    onRetry: options.onRetry ?? (() => void 0),
  };

  let retries = config.retries;
  const factor = config.factor;
  const minTimeout = config.minTimeout;
  const maxTimeout = config.maxTimeout;
  const randomize = config.randomize;
  const retryOn = config.retryOn;
  const onRetry = config.onRetry;

  return new Promise((resolve, reject) => {
    const _fetch = config.fetchFn.bind(
      config.fetchFn,
      ...fetchArguments
    ) as () => Promise<TReturn | TError>;
    const attempt = () => {
      _fetch()
        .then((result) => handleResult(result, resolve))
        .catch((error) => handleResult(error, reject));

      function handleResult(
        result: TReturn | TError,
        resolveOrReject: typeof resolve | typeof reject
      ) {
        if (retryOn(result, retries)) {
          onRetry(result, retries);
          if (retries === 0) {
            reject(result);
            return;
          }

          retries -= 1;

          const timeout = Math.min(
            maxTimeout,
            minTimeout * factor ** (config.retries - retries)
          );

          setTimeout(attempt, randomize ? timeout * Math.random() : timeout);
        } else {
          resolveOrReject(result);
        }
      }
    };

    attempt();
  });
}
